packages/dds/tree/src/feature-libraries/modular-schema/discrepancies.ts

```diff
@@ -346,47 +288,68 @@ function trackObjectNodeDiscrepancies(
 	 * Then, the stored schema is iterated to find the third type.
 	 */
 
-	for (const [fieldKey, fieldStoredSchema] of view.objectNodeFields) {
-		viewFieldKeys.add(fieldKey);
-		if (
-			!stored.objectNodeFields.has(fieldKey) &&
-			fieldStoredSchema.kind !== storedEmptyFieldSchema.kind
-		) {
-			differences.push({
-				identifier: fieldKey,
-				mismatch: "fieldKind",
-				view: fieldStoredSchema.kind,
-				stored: storedEmptyFieldSchema.kind,
-			} satisfies FieldKindDiscrepancy);
-		} else {
-			differences.push(
-				...trackFieldDiscrepancies(
-					view.objectNodeFields.get(fieldKey) as TreeFieldStoredSchema,
-					stored.objectNodeFields.get(fieldKey) as TreeFieldStoredSchema,
-					fieldKey,
-				),
-			);
+	for (const result of compareMaps(view.objectNodeFields, stored.objectNodeFields)) {
+		const fieldKey = result.key;
+		switch (result.type) {
+			case "aExtra": {
+				if (result.value.kind === storedEmptyFieldSchema.kind) {
+					// In one of view/stored, this field is explicitly forbidden, but in the other it is implicitly forbidden
+					// (by way of omission). We treat these identically anyway.
+					break;
+				}
+				yield {
+					identifier: fieldKey,
+					mismatch: "fieldKind",
+					view: result.value.kind,
+					stored: storedEmptyFieldSchema.kind,
+				} satisfies FieldKindDiscrepancy;
+				break;
+			}
+			case "bExtra": {
+				if (result.value.kind === storedEmptyFieldSchema.kind) {
+					// In one of view/stored, this field is explicitly forbidden, but in the other it is implicitly forbidden
+					// (by way of omission). We treat these identically anyway.
+					break;
+				}
+				yield {
+					identifier: fieldKey,
+					mismatch: "fieldKind",
+					view: storedEmptyFieldSchema.kind,
+					stored: result.value.kind,
+				} satisfies FieldKindDiscrepancy;
+				break;
+			}
+			case "both": {
+				yield* getFieldDiscrepancies(result.valueA, result.valueB, fieldKey);
+				break;
+			}
+			default: {
+				break;
+			}
 		}
 	}
+}
 
-	for (const [fieldKey, fieldStoredSchema] of stored.objectNodeFields) {
-		if (viewFieldKeys.has(fieldKey)) {
-			continue;
-		}
-
-		if (fieldStoredSchema.kind !== storedEmptyFieldSchema.kind) {
-			differences.push({
-				identifier: fieldKey,
-				mismatch: "fieldKind",
-				view: storedEmptyFieldSchema.kind,
-				stored: fieldStoredSchema.kind,
-			} satisfies FieldKindDiscrepancy);
+function* compareMaps<K, V1, V2>(
```

Comment:
should this move to utils? it seems generally useful

Reply:
ya, I'll move it in a follow-up